extensions/amp-social-share/1.0/social-share.js

```diff
@@ -0,0 +1,213 @@
+    NAME
+  );
+  const finalEndpoint = createEndpoint(typeConfig, baseEndpoint, props);
+
+  const type = props['type'].toUpperCase();
+  const baseStyle = CSS.BASE_STYLE;
+  const backgroundStyle = CSS[type] || CSS.DEFAULT;
+  const size = {
+    width: checkedWidth,
+    height: checkedHeight,
+  };
+  return (
+    <div
+      role="button"
+      tabindex={props['tabIndex'] || '0'}
+      onKeyDown={(e) => handleKeyPress(e, finalEndpoint)}
+      onClick={() => handleActivation(finalEndpoint)}
+      style={{...size, ...props['style']}}
+      {...props}
+    >
+      <SocialShareIcon
+        style={{...backgroundStyle, ...baseStyle, ...size}}
+        type={type}
+      />
+    </div>
+  );
+}
+
+/**
+ * @param {!JsonObject} props
+ * @param {?string} name
+ * @return {struct}
+ *   {!JsonObject} typeConfig
+ *   {string} baseEndpoint
+ *   {number} checkedWidth
+ *   {number} checkedHeight
+ */
+function checkProps(props, name) {
+  const {
+    'type': type,
+    'shareEndpoint': shareEndpoint,
+    'params': params,
+    'bindings': bindings,
+    'width': width,
+    'height': height,
+  } = props;
+
+  // Verify type is provided
+  if (type === undefined) {
+    throw new Error(`The type attribute is required. ${name}`);
+  }
+
+  // bindings and params props must be objects
+  if (params && !isObject(params)) {
+    throw new Error(`The params property should be an object. ${name}`);
+  }
+  if (bindings && !isObject(bindings)) {
+    throw new Error(`The bindings property should be an object. ${name}`);
+  }
+
+  // User must provide shareEndpoint if they choose a type that is not
+  // pre-configured
+  const typeConfig = getSocialConfig(type) || dict();
+  const baseEndpoint = typeConfig['shareEndpoint'] || shareEndpoint;
+  if (baseEndpoint === undefined) {
+    throw new Error(
+      `A shareEndpoint is required if not using a pre-configured type. ${name}`
+    );
+  }
+
+  const checkedWidth = width ?? DEFAULT_WIDTH;
+  const checkedHeight = height ?? DEFAULT_HEIGHT;
+  return {
+    typeConfig,
+    baseEndpoint,
+    checkedWidth,
+    checkedHeight,
+  };
+}
+
+/**
+ * @param {?string} message
+ */
+function throwWarning(message) {
+  console.warn(message);
+}
+
+/**
+ * @param {!JsonObject} typeConfig
+ * @param {?string} baseEndpoint
+ * @param {!JsonObject} props
+ * @return {?string}
+ */
+function createEndpoint(typeConfig, baseEndpoint, props) {
+  const {'params': params, 'bindings': bindings} = props;
+  const combinedParams = dict({...typeConfig['defaultParams'], ...params});
+  const endpointWithParams = addParamsToUrl(baseEndpoint, combinedParams);
+
+  const combinedBindings = dict();
+  const bindingVars = typeConfig['bindings'];
+  if (bindingVars) {
+    bindingVars.forEach((name) => {
+      combinedBindings[name.toUpperCase()] = combinedParams[name] || '';
+    });
+  }
+  if (bindings) {
+    Object.keys(bindings).forEach((name) => {
+      combinedBindings[name.toUpperCase()] = bindings[name] || '';
```

Comment:
I don't think we intended for any binding to be used here, right?

Reply:
Could you elaborate on your question?  In this section, the bindings object coming from props is being added into the combinedBindings.  Then in the next part we replace variables in finalEndpoint with bindings from combinedBindings.

Reply:
I can pass in `bindings: { foobar: 'x' }`, where `foobar` is some random binding name, and not one of the `typeConfig['bindings']`. I think this is part of the confusion with having `params` and `binding`, we're allowing generic find-and-replace to happen where we meant to just allow you to replace a few strings in our static config JSON with something else.

Reply:
I see your concern.  The API from the original component behaves as you described.

In the previous version of the component, the binding find and replace happens asynchronously (in `urlReplacements.expandUrlAsync()`) and was limited to only those bindings included in our config.

Now the logic is split into two parts, the (synchronous) find and replace logic lives inside the preact component  and we will would have to do an asynchronous call within the amp component (since this needs to be within the amp environment) to get our binding values.  Then pass the bindings into our preact component.

As a side effect to this, someone using the preact component in isolation would be able to add additional generic bindings like you described (`bindings: {foobar: 'x' }`.  I don't see this as a problem though since the user actually gets more flexibility with how they want to use the component and could potentially add other variables specific to the preact environment.

Reply:
I think the model we should try to get to is that the component only has a notion of props/params. And the whole async binding->params mapping is done wholly on the AMP side.

Reply:
Sounds good, let's work toward this model in the AMP component design!